include/dali/core/geom/mat.h

```diff
@@ -0,0 +1,478 @@
+  mat<sub_r, sub_c, Element> result = {};
+  for (size_t j = 0; j < sub_c; j++)
+    for (size_t i = 0; i < sub_r; i++)
+      result(i, j) = m(i+r, j+c);
+  return result;
+}
+
+#define DEFINE_ELEMENTWISE_MAT_MAT_BINOP(op)                            \
+  template <size_t rows, size_t cols, typename T1, typename T2>         \
+  DALI_HOST_DEV inline auto operator op(const mat<rows, cols, T1> &a,   \
+                                        const mat<rows, cols, T2> &b) { \
+    using R = promote_vec_t<T1, T2>;                                    \
+    mat<rows, cols, R> result;                                          \
+    MAT_ELEMENT_LOOP(i, j)                                              \
+      result(i, j) = a(i, j) op b(i, j);                                \
+    return result;                                                      \
+  }
+
+#define DEFINE_ELEMENTWISE_RHS_BINOP(op)                            \
+  template <size_t rows, size_t cols, typename T1, typename T2,     \
+            typename R = promote_vec_scalar_t<T1, T2>>              \
+  DALI_HOST_DEV                                                     \
+  inline std::enable_if_t<is_scalar<T2>::value, mat<rows, cols, R>> \
+    operator op(const mat<rows, cols, T1> &a, const T2 &b) {        \
+    mat<rows, cols, R> result;                                      \
+    MAT_ELEMENT_LOOP(i, j)                                          \
+      result(i, j) = a(i, j) op b;                                  \
+    return result;                                                  \
+  }
+
+#define DEFINE_ELEMENTWISE_LHS_BINOP(op)                            \
+  template <size_t rows, size_t cols, typename T1, typename T2,     \
+            typename R = promote_vec_scalar_t<T2, T1>>              \
+  DALI_HOST_DEV                                                     \
+  inline std::enable_if_t<is_scalar<T1>::value, mat<rows, cols, R>> \
+  operator op(const T1 &a, const mat<rows, cols, T2> &b) {          \
+    mat<rows, cols, R> result;                                      \
+    MAT_ELEMENT_LOOP(i, j)                                          \
+      result(i, j) = a op b(i, j);                                  \
+    return result;                                                  \
+  }
+
+#define DEFINE_ELEMENTWISE_MAT_BINOP(op)\
+  DEFINE_ELEMENTWISE_MAT_MAT_BINOP(op)\
+  DEFINE_ELEMENTWISE_RHS_BINOP(op)\
+  DEFINE_ELEMENTWISE_LHS_BINOP(op)
+
+DEFINE_ELEMENTWISE_MAT_BINOP(+)
+DEFINE_ELEMENTWISE_MAT_BINOP(-)
+DEFINE_ELEMENTWISE_MAT_BINOP(&)
+DEFINE_ELEMENTWISE_MAT_BINOP(|)
+DEFINE_ELEMENTWISE_MAT_BINOP(^)
+DEFINE_ELEMENTWISE_MAT_BINOP(<<)
+DEFINE_ELEMENTWISE_MAT_BINOP(>>)
+
+DEFINE_ELEMENTWISE_LHS_BINOP(*)
+DEFINE_ELEMENTWISE_RHS_BINOP(*)
+DEFINE_ELEMENTWISE_RHS_BINOP(/)
+
+
+template <size_t rows, size_t cols>
+using imat = mat<rows, cols, int>;
+template <size_t rows, size_t cols>
+using dmat = mat<rows, cols, double>;
```

Comment:
Why no `u` and `i16` stuff like the ones for `vec`?

Reply:
I can hardly imagine application of such a matrix. While u8vec3 can represent RGB color, u8mat3 wouldn't suffice for any meaningful transformation thereof.